### ISSUE TITLE
Firebase Admin SDK sends empty message with native image

### ISSUE THREAD

**Issue by veN1337** [GoogleCloudPlatform/native-image-support-java#116]
I'm not quite sure if this issues belongs here or to the [firebase-admin-java repo](https://github.com/firebase/firebase-admin-java), so please let me know and i will open this issue there.

I have a small java app which sends a basic FCM message to a specific device token.
I'm using `firebase-admin:7.1.1` and `google-cloud-graalvm-support:0.3.0` as dependencies and this code snippet to send the message:
```java
FirebaseApp app = FirebaseApp.initializeApp();
Message message = Message.builder()
            .setToken(devicetoken)
            .putData("body", "This is a test message")
            .build();
FirebaseMessaging.getInstance(app).send(message);
```

This works fine when building and running with Java 11.0.10.

But when I build and run a native image with GraalVM Java 11 21.0.0.2 I get a "400 Bad Request" response from the FCM backend (see exception below).

So I took a look at the network traffic and found that the message sent (with the native image) is empty.
I'm quite new to GraalVM and have no idea how to debug or find the cause of this.

#### Request body (token omitted)
```json
{
  "message":{
    "data":{
      "body":"This is a test notification"
    },
    "token":"[...]"
  }
}
```
#### Request headers (Authorization omitted)
```
Accept-Encoding: gzip
Authorization: Bearer [...]
User-Agent: Google-HTTP-Java-Client/1.36.0 (gzip)
x-goog-api-format-version: 2
x-firebase-client: fire-admin-java/7.1.1
Content-Type: application/json; charset=UTF-8
Host: fcm.googleapis.com
Accept: text/html, image/gif, image/jpeg, *; q=.2, */*; q=.2
Connection: keep-alive
Content-Length: 233
```

#### Request body
```json
{
  "message":{}
}
```
#### Request header (Authorization omitted)
```
Accept-Encoding: gzip
Authorization: Bearer [...]
User-Agent: Google-HTTP-Java-Client/1.36.0 (gzip)
x-goog-api-format-version: 2
x-firebase-client: fire-admin-java/7.1.1
Content-Type: application/json; charset=UTF-8
Host: fcm.googleapis.com
Accept: text/html, image/gif, image/jpeg, *; q=.2, */*; q=.2
Connection: keep-alive
Content-Length: 14
```
#### Exception
```java
Exception in thread "main" com.google.firebase.messaging.FirebaseMessagingException: Unexpected HTTP response with status: 400
{
  "error": {
    "code": 400,
    "message": "Recipient of the message is not set.",
    "status": "INVALID_ARGUMENT",
    "details": [
      {
        "@type": "type.googleapis.com/google.rpc.BadRequest",
        "fieldViolations": [
          {
            "field": "message",
            "description": "Recipient of the message is not set."
          }
        ]
      },
      {
        "@type": "type.googleapis.com/google.firebase.fcm.v1.FcmError",
        "errorCode": "INVALID_ARGUMENT"
      }
    ]
  }
}

        at com.google.firebase.messaging.FirebaseMessagingClientImpl$MessagingErrorHandler.createException(FirebaseMessagingClientImpl.java:293)
        at com.google.firebase.messaging.FirebaseMessagingClientImpl$MessagingErrorHandler.createException(FirebaseMessagingClientImpl.java:282)
        at com.google.firebase.internal.AbstractHttpErrorHandler.handleHttpResponseException(AbstractHttpErrorHandler.java:57)
        at com.google.firebase.internal.ErrorHandlingHttpClient.send(ErrorHandlingHttpClient.java:108)
        at com.google.firebase.internal.ErrorHandlingHttpClient.sendAndParse(ErrorHandlingHttpClient.java:72)
        at com.google.firebase.messaging.FirebaseMessagingClientImpl.sendSingleRequest(FirebaseMessagingClientImpl.java:127)
        at com.google.firebase.messaging.FirebaseMessagingClientImpl.send(FirebaseMessagingClientImpl.java:113)
        at com.google.firebase.messaging.FirebaseMessaging$1.execute(FirebaseMessaging.java:137)
        at com.google.firebase.messaging.FirebaseMessaging$1.execute(FirebaseMessaging.java:134)
        at com.google.firebase.internal.CallableOperation.call(CallableOperation.java:36)
        at com.google.firebase.messaging.FirebaseMessaging.send(FirebaseMessaging.java:104)
        at com.google.firebase.messaging.FirebaseMessaging.send(FirebaseMessaging.java:86)
        at fcmnativetest.Application.main(Application.java:23)
Caused by: com.google.api.client.http.HttpResponseException: 400 Bad Request
[response body again (omitted)]
        at com.google.api.client.http.HttpRequest.execute(HttpRequest.java:1113)
        at com.google.firebase.internal.ErrorHandlingHttpClient.send(ErrorHandlingHttpClient.java:96)
        ... 9 more
```

### Site Note
To get the `firebase-admin-sdk` (the message sending) working at all I had to add
```json
{
  "resources": [
    {
      "pattern": "admin_sdk\\.properties"
    }
  ]
}
```
to my `META-INF/native-image/resource-config.json`

**Comment by dzou** [comment]
Thanks for sharing this. Indeed looks like a missing configuration issue on our end.

We just released version `0.4.0` -- Could you try that and see if it fixes the problem? We changed how we registered JSON-related classes for reflection so it might help. 

Let me know if that helps; if not I can try to reproduce on my end. We can also add the `admin_sdk` resource as one of the built-in resources.

**Comment by veN1337** [comment]
Thanks for the quick response. I don't know how I missed the new release 😞 

Unfortunately it didn't fix the problem.

I have pushed a minimal reproducible example here: [veN1337/fcm-native-test](https://github.com/veN1337/fcm-native-test)
The container from the provided Dockerfile needs two environment variables:
-  `GOOGLE_APPLICATION_CREDENTIALS` path to the service account private key
- `FCM_DEVICETOKEN` device token to send the message to

**Comment by dzou** [comment]
Great, thank you for the reproducer. I'll take a look.

**Comment by dzou** [comment]
Hey, I think I might have found a fix. For the temporary solution, I forked your project and added an [additional reflect-config.json](https://github.com/veN1337/fcm-native-test/compare/main...dzou:main#diff-cc783b387f34bc02ac5f61716d44f91c1b614c06c7fab75ff427c71503ef4a4b) which seemed to work for this case.

For the long term solution, I think we will need more time to figure out all of the classes being used reflectively in the library to add this to the project. What makes this tricky is that the reflection used here does not throw error immediately. The firebase library uses reflection to iterate over the fields of an object to populate a JSON request, so if reflection fails, it will simply iterate over an empty list in the first part, and then the error shows up as a "missing request field" later down the line.

There is a good tool to help with debugging called the [Native Image Agent](https://medium.com/graalvm/introducing-the-tracing-agent-simplifying-graalvm-native-image-configuration-c3b56c486271) that we use, maybe this can help you too. In the fork, I ran the app with normal JVM:

```
mvn package
java -agentlib:native-image-agent=config-output-dir=native-image-config -jar target/fcmnativetest-0.1.jar
```

It will generate native-image config files (often generates too much) but it narrows down the choices of what config you need to add.

I'll keep this issue open for now until we add full support for the fire base libs.

**Comment by dzou** [comment]
By the way, is there a way to generate `FCM_DEVICETOKEN`? (Or perhaps is there an emulator we should use?) Not too familiar with Firebase APIs.

I would be interested in adding in a sample application for our native-image tests to support this library.

**Comment by veN1337** [comment]
Thank you so much! The [additional reflect-config.json](https://github.com/veN1337/fcm-native-test/compare/main...dzou:main#diff-cc783b387f34bc02ac5f61716d44f91c1b614c06c7fab75ff427c71503ef4a4b) did the trick and sending messages works now.

Also the [Native Image Agent](https://medium.com/graalvm/introducing-the-tracing-agent-simplifying-graalvm-native-image-configuration-c3b56c486271) you mentioned looks really helpful for (my) future GraalVM problems. Thanks again :)

I'm not quite sure if I understand your last question correctly and I only have very, very basic knowledge of Firebase messaging. If it helps I would be happy to provide a small Android app which, when run in an Android emulator, generates a `FCM_DEVICETOKEN` and receives messages.

**Comment by dzou** [comment]
No problem, glad it works!

Gotcha, when I was debugging your app, I didn't have a valid `FCM_DEVICETOKEN` to pass to the API; so I just made one up, however once I got past your error I started getting an "invalid token error". So I was curious if there was a way to register test/dummy tokens in Firebase or something to let you try calling the APIs and see what happens.

I don't think we are looking for an Android app but rather an sample native-image application which exercises some of those Firebase app calls on a test device token. But don't worry about trying to contribute this; I'm just curious on how to exercise the APIs from an integration testing context.
